### ISSUE TITLE
GuildMember.joinedAt Should Not Be Nullable

### ISSUE THREAD

**Issue by Skillz4Killz** [discordjs/discord.js#3533]
**Please describe the problem you are having in as much detail as possible:**
I was trying to use `member.joinedAt` but typescript said it could be null which led me to try to figure out how this could ever be null. 

```js
    /**
     * The timestamp the member joined the guild at
     * @type {?number}
     */
    this.joinedTimestamp = null;
```

![image](https://user-images.githubusercontent.com/23035000/66670578-1bc12300-ec28-11e9-9379-44bd57fbaf56.png)

![image](https://user-images.githubusercontent.com/23035000/66670583-21b70400-ec28-11e9-95d2-96aed31f97a7.png)

The discord API `joined_at` prop is not optional therefore this should never be null. Discord.js creates a null value for joinedTimestamp in the constructor of the GuildMember.

![image](https://user-images.githubusercontent.com/23035000/66670510-f7654680-ec27-11e9-9035-28de9cfd09e1.png)

**Include a reproducible code sample here, if possible:**
```js
member.joinedAt
```

**Further details:**
- discord.js version: latest master
- Node.js version: 12.0
- Operating system: Mac OSX
- Priority this issue should have – please be realistic and elaborate if possible: Mediumish

- [x] I have also tested the issue on latest master, commit hash:

**Comment by kyranet** [comment]
Is it null even with partials?

**Comment by Skillz4Killz** [comment]
@kyranet Ahh is this cause of partials again. Just another reason to remove partials from discord.js in my opinion. #3486

**Comment by Tenpi** [comment]
I already fixed this in #3493 (null because of partials on master). I'm not sure when it will get merged but hopefully soon ¯\_(ツ)_/¯

On stable, probably an oversight by the devs.

**Comment by iCrawl** [comment]
The reason this is null-able is because of this:

https://github.com/discordjs/discord.js/commit/13bfceb83cf771d1d56d28fd474d824f9a8884db

Completely unrelated to partials.

**Comment by Skillz4Killz** [comment]
Could you please explain why that change was done and if that change is still relevant or needs to be reverted. 

The reason i am confused is why does this line exist the if check seems redundant because it will always exist as per API Docs.
```js
if (data.joined_at) this.joinedTimestamp = new Date(data.joined_at).getTime();
```

**Comment by iCrawl** [comment]
This might come as a surprise now, but Discord API docs sadly are not always accurate.

I'm sure @Lewdcario can follow up with an explanation, and I'm sure its still needed.

**Comment by Lewdcario** [comment]
There is a case where you will receive a `GuildMember` object without a `joined_at` field. To be more specific, in the case you have this structure that isn't cached, and said member comes with a presence update, we will store a new `GuildMember`. It was a required change that happened due to the Discord API (we just noticed it at the time of that commit), and took some time to track down. It is not something we did just for the sake of making a change, that would not be productive, therefore that typing must be marked as nullable.
